tests/models/test_vertexai.py

```diff
@@ -0,0 +1,214 @@
+from __future__ import annotations as _annotations
+
+import json
+from datetime import datetime, timedelta
+from pathlib import Path
+from typing import TYPE_CHECKING, Any
+
+import pytest
+from inline_snapshot import snapshot
+from pytest_mock import MockerFixture
+
+from pydantic_ai import UserError
+from tests.conftest import IsNow
+
+if TYPE_CHECKING:
+    from google.oauth2.service_account import Credentials
+
+    from pydantic_ai.models.vertexai import BearerTokenAuth, VertexAIModel
+
+    google_auth_installed = True
+else:
+    try:
+        # noinspection PyUnresolvedReferences
+        from pydantic_ai.models.vertexai import BearerTokenAuth, VertexAIModel
+    except ImportError:
+        google_auth_installed = False
+    else:
+        google_auth_installed = True
+        from google.oauth2.service_account import Credentials
+
+
+pytestmark = [
+    pytest.mark.skipif(not google_auth_installed, reason='google-auth not installed'),
+    pytest.mark.anyio,
+]
+
+
+async def test_init_service_account(tmp_path: Path, allow_model_requests: None):
+    service_account_path = tmp_path / 'service_account.json'
+    save_service_account(service_account_path, 'my-project-id')
+
+    model = VertexAIModel('gemini-1.5-flash', service_account_file=service_account_path)
+    assert model.url is None
+    assert model.auth is None
+
+    await model.agent_model({}, False, None)
+
+    assert model.url == snapshot(
+        'https://us-central1-aiplatform.googleapis.com/v1/projects/my-project-id/locations/us-central1/'
+        'publishers/google/models/gemini-1.5-flash:'
+    )
+    assert model.auth is not None
+    assert model.name() == snapshot('vertexai:gemini-1.5-flash')
+
+
+class NoOpCredentials:
+    pass
+
+
+async def test_init_env(mocker: MockerFixture, allow_model_requests: None):
+    patch = mocker.patch(
+        'pydantic_ai.models.vertexai.google.auth.default',
+        return_value=(NoOpCredentials(), 'my-project-id'),
+    )
```

Comment:
I tend to prefer passing this stuff in as parameters, e.g. have a parameter called `_auth_provider: AuthProviderProtocol` or something and you pass in a test fake, maybe a mock. You wouldn't be able to do this monkey patching business in Rust 😉

Reply:
I'd rather keen the code clear and mock like this.